packages/internal/src/utils.js

```diff
@@ -411,37 +411,50 @@ export const assertAllDefined = obj => {
   }
 };
 
-const neverDone = harden({ done: false, value: null });
+/** @type {IteratorResult<undefined, never>} */
+const neverDone = harden({ done: false, value: undefined });
+
+/** @type {IteratorResult<never, void>} */
+const alwaysDone = harden({ done: true, value: undefined });
 
-/** @type {AsyncIterable<null>} */
 export const forever = asyncGenerate(() => neverDone);
 
 /**
- * @param {() => unknown} boolFunc
- * `boolFunc`'s return value is used for its truthiness vs falsiness.
+ * @template T
+ * @param {() => T} produce
+ * `produce`'s return value is used for its truthiness vs falsiness.
  * IOW, it is coerced to a boolean so the caller need not bother doing this
  * themselves.
- * @returns {AsyncIterable<null>}
+ * @returns {AsyncIterable<Awaited<T>>}
  */
-export const whileTrue = boolFunc =>
-  asyncGenerate(() =>
-    harden({
-      done: !boolFunc(),
-      value: null,
-    }),
-  );
+export const whileTrue = produce =>
+  asyncGenerate(async () => {
```

Comment:
I looked at `asyncGenerate` to confirm that it does the correct thing if `produce()` throws or returns a promise that rejects. IIUC, it does, in that the async iterator's `next()` will return a promise which rejects. IIRC, this is the correct representation of an async iteration that fails with a reason.

No change suggested. But please let me know if my reasoning is correct.

Reply:
Your reasoning is correct.